### ISSUE TITLE
Find a sensible way to handle many new URI/literals

### ISSUE THREAD

**Issue by pgwillia** [ualbertalib/jupiter#2119]
Let's add a TODO to consider more robust ways of doing this (lookup table in DB or some kind of authority query), because I could see this growing quite large.

_Originally posted by @mbarnett in https://github.com/ualbertalib/jupiter/pull/2089#discussion_r564123624_

Peel will introduce many new URIs in it's metadata.  This is different than the Controlled Vocabularies used before where there were a limited/restricted permitted responses that we drew from for submitting items. The part that is similar is the need to convert to/from URI to humanized string.

**Comment by pgwillia** [comment]
# Problem
Peel will introduce many new URIs in it's metadata. This is different than the Controlled Vocabularies used before where there were a limited/restricted permitted responses that we drew from for submitting items. The part that is similar is the need to convert to/from URI to humanized string. 

I think that having an intermediate symbol is desirable in URI translation so that talking about special cases in our code remain readable.  I.e. `CONTROLLED_VOCABULARIES[:era][:item_type].article`

# What are we doing?

We have a number of [config/controlled_vocabularies](https://github.com/ualbertalib/jupiter/tree/master/config/controlled_vocabularies) which are used to set up the app-wide `CONTROLLED_VOCABULARY` hash.

Each of the URIs in the controlled vocabularies has a symbol which matches the same symbol in our [locales](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/config/locales/en.yml#L776-L831) to give a humanized literal
https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/helpers/application_helper.rb#L4-L13

We use this to 
* [validate that a uri is part of our controlled vocabulary](https://github.com/ualbertalib/jupiter/blob/master/app/validators/uri_validator.rb).
* translate the URI to a human readable [literal for facets](https://github.com/ualbertalib/jupiter/blob/master/app/decorators/facets/uri_value.rb)
* Control Vocab Conversions
  * [licences](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/models/draft_item.rb#L200-L221)
  * [publication status](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/models/draft_item.rb#L223-L231) (and in [paper trail](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/decorators/humanized_change_set.rb#L211-L221))
  * [item types](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/models/draft_item.rb#L233-L259)
  * visibility [item](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/models/draft_item.rb#L261-L294) [thesis](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/models/draft_thesis.rb#L156-L169)
  * language [item](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/models/draft_item.rb#L181-L198) [thesis](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/models/draft_thesis.rb#L138-L154)
  * [institutions](https://github.com/ualbertalib/jupiter/blob/523f6a2aac7e63fade752384222ee632186f3936/app/models/draft_thesis.rb#L171-L188)
* reason about visibilities in [JupiterCore](https://github.com/ualbertalib/jupiter/blob/122fc1a8a40c7b1b4eb312d83424921a97dbab11/app/models/jupiter_core.rb#L12-L16) and [Depositable](https://github.com/ualbertalib/jupiter/blob/122fc1a8a40c7b1b4eb312d83424921a97dbab11/app/models/jupiter_core/depositable.rb#L7-L11)
* create [seeds](https://github.com/ualbertalib/jupiter/blob/master/db/seeds.rb), [fixtures](https://github.com/ualbertalib/jupiter/blob/master/test/fixtures/items.yml), and [tests](https://github.com/ualbertalib/jupiter/tree/master/test)

# What have others done?

## Rails i18n
This guide discusses [i18n in rails](https://guides.rubyonrails.org/i18n.html). 
> easy-to-use and extensible framework for translating your application to a single custom language other than English or for providing multi-language support in your application.

The public API is
```
I18n.t 'store.title' # alias for translate - Lookup text translations
I18n.l Time.now # alias for localize  - Localize Date and Time objects to local formats
```

The default [backend](https://guides.rubyonrails.org/i18n.html#using-different-backends) is [Simple](https://github.com/ruby-i18n/i18n/blob/master/lib/i18n/backend/simple.rb).  There's discussion in the guide about using the [Chain backend](https://github.com/ruby-i18n/i18n/blob/master/lib/i18n/backend/chain.rb).  It's useful when you want to use standard translations with a Simple backend but store custom application translations in a database or other backends.  The same author that worked on i18n gem has a [i18n-active_record](https://github.com/svenfuchs/i18n-active_record) gem that implements an ActiveRecord backend.  This backend stores translations in a Translation table with `locale`, `key`, `value`, `interpolations` and `is_proc` attributes.

Other references
* [Implementing i18n in the database with rails 5](https://medium.com/@toddbaur/implementing-i18n-in-the-database-with-rails-5-98f462d76e72) describes the architecture of having a hybrid locales yml and database for labels.  In the article they also implement a admin user interface to manage translations.
* RubyConf 2018 - Building Generic Software by Chris Salzberg](https://youtu.be/RZkemV_-__A)

## URI Service
[URI Service](https://github.com/cul/uri_service) - found a ruby gem from [Columbia University Libraries](https://library.columbia.edu/)  which is database backed and Solr cached lookup/creation service for URIs.

```
UriService.client.find_term_by_uri('http://id.example.com/123')
# Returns a term hash or nil
#   {
#     'uri' => 'http://id.example.com/123',
#     'value' => 'Batman, John, 1800-1839',
#     'vocabulary_string_key' => 'names',
#     'type' => 'external'
#   }
```

This gem also allows you to 
* create local, external and temporary vocabularies.
* search for a term in a vocabulary `UriService.client.find_terms_by_query('names', 'batman')`

## Linked Data concepts

### [Resource Caching](https://patterns.dataincubator.org/book/resource-caching.html)
> How can an application that relies on loading data be more tolerant of network failures and/or reduce use of bandwidth

### [Label Everything](https://patterns.dataincubator.org/book/label-everything.html)
>How can we ensure that every resource has a basic human-readable name?

[paper](https://link.springer.com/content/pdf/10.1007/978-3-642-25073-6_11.pdf)

**Comment by pgwillia** [comment]
# Solution

Generate an ActiveRecord Model called `Vocabulary` with `namespace`, `vocab`, `uri`, and `code` attributes.  Perhaps make the `code` optional and derive it from snake case of the `label`.

The API for this model:
* `Vocabulary.<namespace>.<vocab>.<code>` returns a `uri`
* `Vocabulary.<namespace>.<vocab>.from_uri(uri)` returns a `code`
* `t('vocabulary.<namespace>.<vocab>.<code>')` returns a label

Use the ActiveRecord backend for i18n to store the translations/labels for each URI/code.  Have a hook on create for Vocabulary which adds the translation to the i18n ActiveRecord backend.

## Other considerations

* Look first in Vocabulary but fall back to our CONTROLLED_VOCABULARIES if not found.

**Comment by pgwillia** [comment]
This was an exercise to see if we could get a label from a URI programmatically.   We might do this once on ingest and then store in our database(s).

[linkeddata](https://rubygems.org/gems/linkeddata)
```
require 'rdf'
require 'linkeddata'

graph = RDF::Graph.load("http://id.loc.gov/authorities/names/n79007225") # a URI from the FolkFest data
graph.query({predicate: RDF::Vocab::SKOS.prefLabel}).first.object.to_s
=> "Edmonton (Alta.)"

graph.query({predicate: RDF::Vocab::SKOS.prefLabel}) { |statement| puts statement.object.to_s }
Edmonton (Alta.)
Alberta--Edmonton Region
Alberta--Edmonton Metropolitan Area
Alberta--Edmonton
Alberta--Edmonton Suburban Area
Edmonton
Strathcona (Edmonton, Alta.)--(DLC)n 81057372
```

From [Getting data from the Semantic Web (Ruby)](http://web.archive.org/web/20190222010929/https://semanticweb.org/wiki/Getting_data_from_the_Semantic_Web_(Ruby)) and [Queryable:query](https://www.rubydoc.info/github/ruby-rdf/rdf/RDF%2FQueryable:query)

**Comment by mbarnett** [comment]
Pretty sure there's a much cheaper way to do that, because the RDF::Graph calls are *very* expensive/slow. This was a huge issue in ActiveTriples/ActiveFedora. Give me a bit to check

**Comment by mbarnett** [comment]
Hmmm, so you can do something like:

```ruby
irb(main):016:0> uri = "http://purl.org/dc/terms/title"
=> "http://purl.org/dc/terms/title"
irb(main):017:0> RDF::Vocabulary.find_term(uri)
=> #<RDF::Vocabulary::Term:0x3fcd9556c1a4 ID:http://purl.org/dc/terms/title>
irb(main):018:0> RDF::Vocabulary.find(uri)
=> RDF::StrictVocabulary(http://purl.org/dc/terms/)
irb(main):019:0> RDF::Vocabulary.find_term(uri).label.to_s
=> "Title"
```

generally, which is good both because it bypasses all the in memory graph stuff and because in your example we have to know the vocab the URI comes from to form the query, which is a bunch of extra work to have to do if we can't just have a Vocab-agnostic conversion. It's still not terribly efficient because it does this in just about the dumbest way possible – it simply looks through every vocab it knows about sequentially trying to find the corresponding URI. But the Graph query is worse in that I think it's built on top of this, so it's several more layers of inefficient on top of this inefficiency.

Looking up the specific URI "http://id.loc.gov/authorities/names/n79007225" doesn't seem to work, however. Could be a bug in RDF, could be something else at play there.

Long story short, `humanize_uri` gets called a lot and it's in a hot path on the facets rendering code, so it can't be slow without having a very large impact on performance. That probably rules out anything like the Graph approach, the RDF find methods, and definitely querying external authorities whose labels we may not even like anyways. I think we're going to have to store the mappings in the DB by hand and only rely on this kind of method for an initial population of the values, and only where the labels these methods provide are sufficiently user-friendly to be worth using.

Not sure there's any value in the URI Service gem either beyond what we could more easily do ourselves without having to worry that it's abandoned.

**Comment by pgwillia** [comment]
> Looking up the specific URI "http://id.loc.gov/authorities/names/n79007225" doesn't seem to work, however. Could be a bug in RDF, could be something else at play there.

There could be something else at play there.  The example you gave was the URI corresponding to a predicate.  The example I'm trying to solve is the object as a URI.  I tried some of the other URIs in our [config/controlled_vocabularies](https://github.com/ualbertalib/jupiter/tree/master/config/controlled_vocabularies) and none of these worked either.

This is an example that didn't work with RDF::Graph so this method isn't bullet proof either.
https://github.com/ualbertalib/jupiter/blob/64b67865e18f79c9463d4caf127393ff14b65fc4/config/controlled_vocabularies/digitization_language.yml#L2

```
ERROR <https://iso639-3.sil.org/code/eng>: 5:501: FATAL: error parsing attribute name
5:501: FATAL: attributes construct error
5:501: FATAL: Couldn't find end of Start Tag n.length line 5
5:780: FATAL: xmlParseEntityRef: no name
5:781: FATAL: xmlParseEntityRef: no name
5:1599: FATAL: xmlParseEntityRef: no name
5:1600: FATAL: xmlParseEntityRef: no name
5:1697: FATAL: xmlParseEntityRef: no name
5:1713: FATAL: EntityRef: expecting ';'
5:2086: FATAL: xmlParseEntityRef: no name
5:2088: FATAL: EntityRef: expecting ';'
5:2181: FATAL: xmlParseEntityRef: no name
5:2183: FATAL: EntityRef: expecting ';'
5:2321: FATAL: xmlParseEntityRef: no name
5:2322: FATAL: xmlParseEntityRef: no name
5:2374: FATAL: StartTag: invalid element name
5:2410: FATAL: xmlParseEntityRef: no name
5:2412: FATAL: EntityRef: expecting ';'
5:2413: FATAL: StartTag: invalid element name
5:2554: FATAL: xmlParseEntityRef: no name
5:2555: FATAL: xmlParseEntityRef: no name
5:2587: FATAL: xmlParseEntityRef: no name
5:2588: FATAL: xmlParseEntityRef: no name
5:2628: FATAL: xmlParseEntityRef: no name
5:2664: FATAL: EntityRef: expecting ';'
5:2764: FATAL: xmlParseEntityRef: no name
5:2765: FATAL: xmlParseEntityRef: no name
5:3397: FATAL: error parsing attribute name
5:3397: FATAL: attributes construct error
5:3397: FATAL: Couldn't find end of Start Tag r.length line 5
5:3546: FATAL: xmlParseEntityRef: no name
5:3556: FATAL: EntityRef: expecting ';'
5:3572: FATAL: xmlParseEntityRef: no name
5:3582: FATAL: EntityRef: expecting ';'
5:3600: FATAL: xmlParseEntityRef: no name
5:3601: FATAL: xmlParseEntityRef: no name
5:3714: FATAL: xmlParseEntityRef: no name
5:3723: FATAL: EntityRef: expecting ';'
5:3801: FATAL: xmlParseEntityRef: no name
5:3803: FATAL: EntityRef: expecting ';'
5:3804: FATAL: xmlParseEntityRef: no name
5:3830: FATAL: EntityRef: expecting ';'
5:4069: FATAL: xmlParseEntityRef: no name
5:4070: FATAL: xmlParseEntityRef: no name
5:4223: FATAL: xmlParseEntityRef: no name
5:4224: FATAL: xmlParseEntityRef: no name
5:4378: FATAL: xmlParseEntityRef: no name
5:4379: FATAL: xmlParseEntityRef: no name
5:4428: FATAL: StartTag: invalid element name
5:4440: FATAL: error parsing attribute name
5:4440: FATAL: attributes construct error
5:4440: FATAL: Couldn't find end of Start Tag i line 5
5:4556: FATAL: xmlParseEntityRef: no name
5:4582: FATAL: EntityRef: expecting ';'
5:4583: FATAL: xmlParseEntityRef: no name
5:4584: FATAL: xmlParseEntityRef: no name
5:4723: FATAL: xmlParseEntityRef: no name
5:4725: FATAL: EntityRef: expecting ';'
5:4788: FATAL: xmlParseEntityRef: no name
5:4789: FATAL: xmlParseEntityRef: no name
5:4813: FATAL: xmlParseEntityRef: no name
5:4815: FATAL: EntityRef: expecting ';'
5:4816: FATAL: xmlParseEntityRef: no name
5:4818: FATAL: EntityRef: expecting ';'
5:4866: FATAL: error parsing attribute name
5:4866: FATAL: attributes construct error
5:4866: FATAL: Couldn't find end of Start Tag u line 5
5:4910: FATAL: xmlParseEntityRef: no name
5:4917: FATAL: EntityRef: expecting ';'
5:5022: FATAL: error parsing attribute name
5:5022: FATAL: attributes construct error
5:5022: FATAL: Couldn't find end of Start Tag n.length line 5
5:5036: FATAL: xmlParseEntityRef: no name
5:5045: FATAL: EntityRef: expecting ';'
5:5418: FATAL: xmlParseEntityRef: no name
5:5419: FATAL: xmlParseEntityRef: no name
5:5658: FATAL: xmlParseEntityRef: no name
5:5670: FATAL: EntityRef: expecting ';'
5:6026: FATAL: xmlParseEntityRef: no name
5:6027: FATAL: xmlParseEntityRef: no name
5:6273: FATAL: xmlParseEntityRef: no name
5:6286: FATAL: EntityRef: expecting ';'
5:6287: FATAL: xmlParseEntityRef: no name
5:6303: FATAL: EntityRef: expecting ';'
5:6304: FATAL: xmlParseEntityRef: no name
5:6306: FATAL: EntityRef: expecting ';'
5:6564: FATAL: xmlParseEntityRef: no name
5:6566: FATAL: EntityRef: expecting ';'
5:6762: FATAL: xmlParseEntityRef: no name
5:6775: ERROR: Entity 'v.prototype' not defined
5:7025: FATAL: xmlParseEntityRef: no name
5:7027: FATAL: EntityRef: expecting ';'
5:7028: FATAL: xmlParseEntityRef: no name
5:7030: FATAL: EntityRef: expecting ';'
5:7034: FATAL: xmlParseEntityRef: no name
5:7035: FATAL: xmlParseEntityRef: no name
5:7816: FATAL: error parsing attribute name
5:7816: FATAL: attributes construct error
5:7816: FATAL: Couldn't find end of Start Tag t.length line 5
5:8110: FATAL: xmlParseEntityRef: no name
5:8122: FATAL: EntityRef: expecting ';'
5:8316: FATAL: xmlParseEntityRef: no name
5:8317: FATAL: xmlParseEntityRef: no name
5:8362: FATAL: xmlParseEntityRef: no name
5:8364: FATAL: EntityRef: expecting ';'
5:8385: FATAL: xmlParseEntityRef: no name
5:8393: FATAL: EntityRef: expecting ';'
5:8394: FATAL: xmlParseEntityRef: no name
5:8395: FATAL: xmlParseEntityRef: no name
5:8557: FATAL: error parsing attribute name
5:8557: FATAL: attributes construct error
5:8557: FATAL: Couldn't find end of Start Tag e line 5
36:8: FATAL: Opening and ending tag mismatch: script line 25 and head
501:8: FATAL: Opening and ending tag mismatch: meta line 6 and html
502:1: FATAL: EndTag: '</' not found
```

> Long story short, `humanize_uri` gets called a lot and it's in a hot path on the facets rendering code, so it can't be slow without having a very large impact on performance. That probably rules out anything like the Graph approach, the RDF find methods, and definitely querying external authorities whose labels we may not even like anyways. I think we're going to have to store the mappings in the DB by hand and only rely on this kind of method for an initial population of the values, and only where the labels these methods provide are sufficiently user-friendly to be worth using.

I have no intention of using RDF::Graph in the application code for any just in time type lookup.  The problem I was trying to solve with that snippet was looking up a reasonable label to the URIs in the Folk Fest data.  I did this by hand for [config/controlled_vocabularies/digitization_*] in [my FolkFest Modelling PR](https://github.com/ualbertalib/jupiter/pull/2089) by visiting the URI in my browser to find a reasonable `code`.  I hadn't yet even done the work to fill in the values for the i18n translations.  When we do load the metadata we could use something like that snippet with RDF::Graph in an ActiveJob to lookup and store an appropriate label.  Interpreting URIs is one of the activities that I didn't think would scale when we add more digitization content.  The label would then be stored in the i18n ActiveRecord backend and used by humanize_uri whenever the app needs it.

> Not sure there's any value in the URI Service gem either beyond what we could more easily do ourselves without having to worry that it's abandoned.
💯  agree.  The URI Service gem will not be useful for us. This was just part of my scan of the state of the world.

**Comment by mbarnett** [comment]
Hmmm yeah good point, wasn't thinking in terms of Predicate vs Object.

I almost want to say we should look at more of a Questioning Authority approach for populating the DB, but I'm still concerned about some of the labels maybe not being what we'd *want* to present to an end user (perhaps needlessly worried, but Hydra was so bad for presenting really poor UX due to that kind of reliance on mechanical mapping)

**Comment by pgwillia** [comment]
> generally, which is good both because it bypasses all the in memory graph stuff and because in your example we have to know the vocab the URI comes from to form the query, which is a bunch of extra work to have to do if we can't just have a Vocab-agnostic conversion. It's still not terribly efficient because it does this in just about the dumbest way possible – it simply looks through every vocab it knows about sequentially trying to find the corresponding URI. But the Graph query is worse in that I think it's built on top of this, so it's several more layers of inefficient on top of this inefficiency.

It would be nice to have something where we could just call `label` and have it do the right thing.  That's where I did some reading about "Linked Data Concepts" and how this plays out in most graphs.  There are some other predicates to look at if `RDF::Vocab::SKOS.prefLabel` isn't there like `skos:altLabel` and `rdfs:label`.  From what I gather `prefLabel` should be vocabulary agnostic.  We're trying to find the edges of the graph not the connecting parts, if that makes any sense.

**Comment by pgwillia** [comment]
Tell me more about Questioning Authority approach?

The blog post about the i18n ActiveRecord backend includes a basic admin interface for adding/editing translations.  Would that help with fixing bad mappings?

**Comment by mbarnett** [comment]
I was actually thinking less in terms of "it might not be `prefLabel`" and more in terms of "it might not be in `RDF::Vocab::SKOS`" but yeah, there might be *two* ways for that to go wrong.

Lemme mull the questioning authority vs graph query approach some more, may or may not come up with some thoughts

**Comment by pgwillia** [comment]
First we can create a graph from a spreadsheet like 'FolkFestTriples - v.1' downloaded as a csv called 'follkfest.csv' in the current path.
```
graph = RDF::Graph.new
CSV.foreach('folkfest.csv', headers: true) do |row|
  subject = RDF::URI.new(row["Entity"], validate: true)
  predicate = RDF::URI.new(row["Property"], validate: true)
  object = begin
    RDF::URI.new(row["Value"], validate: true) 
  rescue
    RDF::Literal.new(row["Value"])
  end
  graph << [subject, predicate, object]
end
```

Then we can find the URI's and their labels
```
query = RDF::Query.new do
  pattern [:program, :predicate, :term]
  pattern [:term, RDF::Vocab::SKOS.prefLabel, :label]
end
graph.query(query).select(:predicate, :term, :label).distinct { |statement| puts statement.inspect }
```

predicate | term | label
-- | -- | --
<http://purl.org/dc/elements/1.1/coverage> | <http://id.loc.gov/authorities/names/n79007225> | "Edmonton (Alta.)"
<http://purl.org/dc/elements/1.1/subject> | <http://id.loc.gov/authorities/subjects/sh2008004058> | "Folk music festivals"
<http://purl.org/dc/terms/type> | <http://id.loc.gov/vocabulary/resourceTypes/txt> | "Text"
<http://www.europeana.eu/schemas/edm/hasType> | <http://id.loc.gov/authorities/genreForms/gf2014026156> | "Programs (Publications)"
<http://id.loc.gov/vocabulary/relators/pup> | <http://id.loc.gov/authorities/names/n79007225> | "Edmonton (Alta.)"
<http://purl.org/dc/terms/language> | <https://iso639-3.sil.org/code/eng> | "English"
<http://www.europeana.eu/schemas/edm/rights> | <https://rightsstatements.org/page/InC/1.0/?language=en> | "In Copyright

**Comment by pgwillia** [comment]
I think this is solved by @mbarnett's ControlledVocabulary API
